### PR TITLE
[StructuralHash] Global Variable

### PR DESCRIPTION
**Pull request by kyulee-com**
This update enhances the implementation of structural hashing for global variables, using their initial contents. Private global variables or constants are often used for metadata, where their names are not unique. This can lead to the creation of different hash results although they could be merged by the linker as they are effectively identical.
  - Refine the hashing of GlobalVariables for strings or certain Objective-C metadata cases that have section names. This can be further extended to other scenarios.
  - Expose StructuralHash for GlobalVariable so that this API can be utilized by MachineStableHashing, which is also employed in the global function outliner.

This change significantly improves size reduction by an additional 1% on the LLD binary when the global function outliner and merger are enabled together. As discussed in the RFC https://discourse.llvm.org/t/loh-conflicting-with-machineoutliner/83279/8?u=kyulee-com, if we disable or relocate the LOH pass, the size impact could increase to 4%.

### CODE REVIEW

**Comment by kyulee-com** [comment]
cc @nocchijiang

File: llvm/lib/IR/StructuralHash.cpp [hunk 1]
```diff
@@ -65,19 +65,44 @@ class StructuralHashImpl {
...
   }
 
-  stable_hash hashGlobalValue(const GlobalValue *GV) {
+  static stable_hash hashGlobalVariable(const GlobalVariable &GVar) {
+    if (!GVar.hasInitializer())
+      return hashGlobalValue(&GVar);
+
+    // Hash the contents of a string.
+    if (GVar.getName().starts_with(".str"))
+      return hashConstant(GVar.getInitializer());
```
[the remark this answers is not in the document]
Comment (kyulee-com) [reply llvm/lib/IR/StructuralHash.cpp:86]: I only use `hashInitializer()` to determine if the global variable (beginning with `.str`) has an initializer. While I don't specifically verify if it's a string literal, this approach should suffice for most scenarios. It's still safe for our purposes, even if it's not a string literal. The only concern is avoiding an initializer that recursively references the current global variable itself. I assume that the cases filtered out here and below do not include such instances.

File: llvm/lib/IR/StructuralHash.cpp [hunk 2]
```diff
@@ -65,19 +65,44 @@ class StructuralHashImpl {
...
+    static constexpr const char *SectionNames[] = {
+        "__cfstring",      "__cstring",      "__objc_classrefs",
+        "__objc_methname", "__objc_selrefs",
+    };
+    if (GVar.hasSection()) {
+      StringRef SectionName = GVar.getSection();
+      for (const char *Name : SectionNames) {
+        if (SectionName.contains(Name))
+          return hashConstant(GVar.getInitializer());
+      }
```
Comment (ellishg) [inline llvm/lib/IR/StructuralHash.cpp:99]:
```suggestion
      for (const char *Name : SectionNames)
        if (SectionName.contains(Name))
          return hashConstant(GVar.getInitializer());
```

File: llvm/lib/IR/StructuralHash.cpp [hunk 3]
```diff
@@ -65,19 +65,44 @@ class StructuralHashImpl {
...
+    // Hash structural contents of Objective-C metadata in specific sections.
+    // This can be extended to other metadata if needed.
+    static constexpr const char *SectionNames[] = {
+        "__cfstring",      "__cstring",      "__objc_classrefs",
+        "__objc_methname", "__objc_selrefs",
+    };
+    if (GVar.hasSection()) {
+      StringRef SectionName = GVar.getSection();
+      for (const char *Name : SectionNames) {
+        if (SectionName.contains(Name))
```
Comment (ellishg) [inline llvm/lib/IR/StructuralHash.cpp:97]: Would it be correct to use `ends_with()`?

File: llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+; RUN: llc -mtriple=arm64-apple-darwin -enable-global-merge-func=true -global-merging-skip-no-params=false < %t/ns-const.ll | FileCheck %s
+; RUN: llc -mtriple=arm64-apple-darwin -enable-global-merge-func=true -global-merging-skip-no-params=false < %t/objc-ref.ll | FileCheck %s
+
+; CHECK: _f1.Tgm
+; CHECK: _f2.Tgm
+
+;--- string.ll
+
+@.str = private unnamed_addr constant [6 x i8] c"hello\00", align 1
+@.str.1 = private unnamed_addr constant [6 x i8] c"hello\00", align 1
```
Comment (ellishg) [inline llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll:17]: Can you add another global string with a different value to ensure it isn't merged?

File: llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll [hunk 5]
```diff
@@ -0,0 +1,91 @@
+; This test verifies that global variables are hashed based on their initial contents,
+; allowing them to be merged even if they appear different due to their names.
+; Now they become identical functions that can be merged without creating a paramter.
+
+; RUN: rm -rf %t && split-file %s %t
+
+; RUN: llc -mtriple=arm64-apple-darwin -enable-global-merge-func=true -global-merging-skip-no-params=false < %t/string.ll | FileCheck %s
+; RUN: llc -mtriple=arm64-apple-darwin -enable-global-merge-func=true -global-merging-skip-no-params=false < %t/ns-const.ll | FileCheck %s
+; RUN: llc -mtriple=arm64-apple-darwin -enable-global-merge-func=true -global-merging-skip-no-params=false < %t/objc-ref.ll | FileCheck %s
```
Comment (ellishg) [inline llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll:9]: These seem like three separate tests that don't need to share any data (except that their `CHECK` lines happen to be the same). Can we separate them into their own files?

File: llvm/test/CodeGen/AArch64/cgdata-outline-gvar.ll [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+; RUN: llvm-objdump -d %t_write_base | FileCheck %s
+
+; RUN: llvm-cgdata --merge %t_write_base -o %t_cgdata_base
+
+; Read the cgdata in the machine outliner for optimistically outlining in local-one.ll.
+; RUN: llc -mtriple=arm64-apple-darwin -enable-machine-outliner -codegen-data-use-path=%t_cgdata_base  -aarch64-enable-collect-loh=false -append-content-hash-outlined-name=false -filetype=obj %t/local-one.ll -o %t_read_base
+; RUN: llvm-objdump -d %t_read_base | FileCheck %s
+
+; The names of globals `.str` and `.str.4` are different, but their initial contents are identical.
+; The outlined function now starts with a reference to that global ("hello\00").
```
Comment (ellishg) [inline llvm/test/CodeGen/AArch64/cgdata-outline-gvar.ll:17]: I'm a bit confused. Where is the reference to the "hello\00" string in this outlined function?

File: llvm/lib/IR/StructuralHash.cpp [hunk 7]
```diff
@@ -65,19 +65,44 @@ class StructuralHashImpl {
...
     return hashAPInt(F.bitcastToAPInt());
   }
 
-  stable_hash hashGlobalValue(const GlobalValue *GV) {
+  static stable_hash hashGlobalVariable(const GlobalVariable &GVar) {
+    if (!GVar.hasInitializer())
+      return hashGlobalValue(&GVar);
+
+    // Hash the contents of a string.
+    if (GVar.getName().starts_with(".str"))
```
Comment (ellishg) [inline llvm/lib/IR/StructuralHash.cpp:85]: This seems like a weird way to detect if a GV is a string. Is there a better way? Could we check if the type if `[* x i8]` or something?

File: llvm/lib/IR/StructuralHash.cpp [hunk 8]
```diff
@@ -98,14 +123,21 @@ class StructuralHashImpl {
...
+
     if (auto *G = dyn_cast<GlobalValue>(C)) {
       Hashes.emplace_back(hashGlobalValue(G));
       return stable_hash_combine(Hashes);
     }
 
     if (const auto *Seq = dyn_cast<ConstantDataSequential>(C)) {
-      Hashes.emplace_back(xxh3_64bits(Seq->getRawDataValues()));
-      return stable_hash_combine(Hashes);
+      if (Seq->isString()) {
```
Comment (nocchijiang) [inline llvm/lib/IR/StructuralHash.cpp:137]: What about other `ConstantDataSequential`s which are not string?

File: llvm/lib/CodeGen/MachineStableHash.cpp [hunk 9]
```diff
@@ -97,9 +99,14 @@ stable_hash llvm::stableHashValue(const MachineOperand &MO) {
       ++StableHashBailingGlobalAddress;
       return 0;
     }
-    auto Name = GV->getName();
-    return stable_hash_combine(MO.getType(), MO.getTargetFlags(),
-                               stable_hash_name(Name), MO.getOffset());
+    stable_hash GVHash = 0;
+    if (auto *GVar = dyn_cast<GlobalVariable>(GV))
```
Comment (nocchijiang) [inline llvm/lib/CodeGen/MachineStableHash.cpp:103]: Seems like we can generate meaningful hash for globals regardless of the presence of name. Should we move the `hasName` check after this?

File: llvm/lib/IR/StructuralHash.cpp (hunk 7 above)
Comment (kyulee-com) [reply llvm/lib/IR/StructuralHash.cpp]: Added a string condition, using `ConstantDataSequential`.

File: llvm/test/CodeGen/AArch64/cgdata-outline-gvar.ll (hunk 6 above)
Comment (kyulee-com) [reply llvm/test/CodeGen/AArch64/cgdata-outline-gvar.ll]: Restructured the test cases.

File: llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll (hunk 5 above)
Comment (kyulee-com) [reply llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll]: Separated the test cases into their own files.

File: llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll (hunk 4 above)
Comment (kyulee-com) [reply llvm/test/CodeGen/AArch64/cgdata-merge-gvar.ll]: Added a failing case.

File: llvm/lib/IR/StructuralHash.cpp (hunk 3 above)
Comment (kyulee-com) [reply llvm/lib/IR/StructuralHash.cpp:101]: Unfortunately, we can't as the section name can be in the middle. You see an example, `cgdata-merge-gvar-objc.ll`, like `section "__DATA,__objc_selrefs,literal_pointers,no_dead_strip"`.

File: llvm/lib/IR/StructuralHash.cpp (hunk 8 above)
Comment (kyulee-com) [reply llvm/lib/IR/StructuralHash.cpp:140]: If you see below in the switch statement, `ConstantArrayVal` or `ConstantVectorVal` (that inherits `ConstantDataSequentials`) should cover the rest by handling each Constant element.

File: llvm/lib/CodeGen/MachineStableHash.cpp (hunk 9 above)
Comment (kyulee-com) [reply llvm/lib/CodeGen/MachineStableHash.cpp:99]: Moved `hasName` check below.

**Comment by kyulee-com** [comment]
@mingmingl-llvm @nocchijiang @ellishg I think I've addressed all the comments. Could you please review it again?"

**Review by thevinster — APPROVED** [review]
lgtm

**Pull request merged**